## Decision Trees
### Contents
1. Decision Tree for Classification
2. Decision Tree for Regression
3. Advantages & Limitations of Decision Trees

* Importing the required Libraries

In [141]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
# %matplotlib inline

In [142]:
play_data = pd.read_csv('Data/tennis.csv.txt')

In [143]:
play_data

outlook  temp humidity  windy play
0      sunny   hot     high  False   no
1      sunny   hot     high   True   no
2   overcast   hot     high  False  yes
3      rainy  mild     high  False  yes
4      rainy  cool   normal  False  yes
5      rainy  cool   normal   True   no
6   overcast  cool   normal   True  yes
7      sunny  mild     high  False   no
8      sunny  cool   normal  False  yes
9      rainy  mild   normal  False  yes
10     sunny  mild   normal   True  yes
11  overcast  mild     high   True  yes
12  overcast   hot   normal  False  yes
13     rainy  mild     high   True   no

### Entropy of play

* p is the probability ( How many Yes or No out of How many ?? )
* Entropy(play) =  – p(Yes) . log2p(Yes) – p(No) . log2p(No) and SO ON if more classes are there

In [144]:
play_data.play.value_counts()

yes    9
no     5
Name: play, dtype: int64

In [145]:
Entropy_Play = -(9/14)*np.log2(9/14) -(5/14)*np.log2(5/14)

In [146]:
Entropy_Play

0.9402859586706311

In [147]:
play_data.dtypes

outlook     object
temp        object
humidity    object
windy         bool
play        object
dtype: object

### Information Gain
* Gain(S, A) = Entropy(S) – ∑ [ p(S|A) . Entropy(S|A) ]
* We intend to choose the attribute, splitting by which information gain will be the most

#### Information Gain on splitting by Outlook
* Gain(Play, Outlook) = Entropy(Play) – ∑ [ p(Play|Outlook) . Entropy(Play|Outlook) ]   --> Regular Formula 
* 
* By Expanding we get.
* 
* Gain(Play, Outlook) = Entropy(Play) – [ p(Play|Outlook=Sunny) . Entropy(Play|Outlook=Sunny) ] – [ p(Play|Outlook=Overcast) . Entropy(Play|Outlook=Overcast) ]  – [ p(Play|Outlook=Rain) . Entropy(Play|Outlook=Rain) ] 

In [148]:
play_data[play_data.outlook == 'sunny']

outlook  temp humidity  windy play
0    sunny   hot     high  False   no
1    sunny   hot     high   True   no
7    sunny  mild     high  False   no
8    sunny  cool   normal  False  yes
10   sunny  mild   normal   True  yes

In [149]:
# Entropy(Play|Outlook=Sunny)
Entropy_Play_Outlook_Sunny =-(3/5)*np.log2(3/5) -(2/5)*np.log2(2/5)

In [150]:
Entropy_Play_Outlook_Sunny

0.9709505944546686

In [ ]:
##############

In [151]:
play_data[play_data.outlook == 'overcast']

outlook  temp humidity  windy play
2   overcast   hot     high  False  yes
6   overcast  cool   normal   True  yes
11  overcast  mild     high   True  yes
12  overcast   hot   normal  False  yes

In [152]:
# Entropy(Play|Outlook=overcast)
Entropy_Play_Outlook_Overcast = -(4/4)*np.log2(4/4) - (4/4)*np.log2(4/4)

In [153]:
Entropy_Play_Outlook_Overcast

-0.0

In [ ]:
#################

In [ ]:
# Entropy(Play|Outlook=Overcast)
# Since, it's a homogenous data entropy will be 0

In [154]:
play_data[play_data.outlook == 'rainy']

outlook  temp humidity  windy play
3    rainy  mild     high  False  yes
4    rainy  cool   normal  False  yes
5    rainy  cool   normal   True   no
9    rainy  mild   normal  False  yes
13   rainy  mild     high   True   no

In [155]:
# Entropy(Play|Outlook=rainy)
Entropy_Play_Outlook_Rain = -(2/5)*np.log2(2/5) - (3/5)*np.log2(3/5)

In [156]:
Entropy_Play_Outlook_Rain

0.9709505944546686

#### Gain on splitting by attribute " outlook " - We need to calculate the same for other factors also to compare the Gain in order to get the best split. 

In [157]:
#Gain(Play, Outlook) = Entropy(Play) – [ p(Play|Outlook=Sunny) . Entropy(Play|Outlook=Sunny) ] – 
#[ p(Play|Outlook=Overcast) . Entropy(Play|Outlook=Overcast) ] – [ p(Play|Outlook=Rain) . Entropy(Play|Outlook=Rain) ]

Entropy_Play - (5/14)*Entropy_Play_Outlook_Sunny - (4/14)*0 - (5/14) * Entropy_Play_Outlook_Rain 

0.24674981977443933

#### Other gains - Calculated using the similar steps as above
* Gain(Play, Outlook) - 0.2467
* Gain(Play, Temperature) - 0.029
* Gain(Play, Humidity) - 0.151
* Gain(Play, Wind) - 0.048

#### Conclusion - Outlook is winner & thus becomes root of the tree

In [ ]:
######################################################################################################

### Time to find the next splitting criteria - " Outlook " is the root now for this Sub tree / Branch

In [159]:
play_data[play_data.outlook == 'overcast']

outlook  temp humidity  windy play
2   overcast   hot     high  False  yes
6   overcast  cool   normal   True  yes
11  overcast  mild     high   True  yes
12  overcast   hot   normal  False  yes

##### Conclusion - If outlook is overcast, play is true

### Let's find the next splitting feature

In [160]:
play_data[play_data.outlook == 'sunny']

outlook  temp humidity  windy play
0    sunny   hot     high  False   no
1    sunny   hot     high   True   no
7    sunny  mild     high  False   no
8    sunny  cool   normal  False  yes
10   sunny  mild   normal   True  yes

In [161]:
# Entropy(Play_Sunny|)
Entropy_Play_Outlook_Sunny =-(3/5)*np.log2(3/5) -(2/5)*np.log2(2/5)

In [162]:
Entropy_Play_Outlook_Sunny

0.9709505944546686

### Information Gain for humidity

In [163]:
#Entropy for attribute high = 0, also entropy for attribute normal = 0 
Entropy_Play_Outlook_Sunny - (3/5)*0 - (2/5)*0

0.9709505944546686

In [164]:
np.log2(2/2)

0.0

In [ ]:
# Same will be for attribute = high

### Information Gain for windy
* False -> 3 -> [1+ 2-]
* True -> 2 -> [1+ 1-]

In [165]:
Entropy_Wind_False = -(1/3)*np.log2(1/3) - (2/3)*np.log2(2/3)

In [166]:
Entropy_Wind_False

0.9182958340544896

In [167]:
Entropy_Play_Outlook_Sunny - (3/5)* Entropy_Wind_False  - (2/5)*1 

0.01997309402197489

In [ ]:
# entropy - 1

### Information Gain for temperature
* hot -> 2 -> [2- 0+]
* mild -> 2 -> [1+ 1-]
* cool -> 1  -> [1+ 0-]

In [168]:
Entropy_Play_Outlook_Sunny - (2/5)*0 - (1/5)*0 - (2/5)* 1

0.5709505944546686

#### Conclusion : Humidity is the best choice on sunny branch

<img src="https://github.com/awantik/machine-learning-slides/blob/master/dt3.PNG?raw=true" width="600px">

In [169]:
play_data[(play_data.outlook == 'sunny') & (play_data.humidity == 'high')]

outlook  temp humidity  windy play
0   sunny   hot     high  False   no
1   sunny   hot     high   True   no
7   sunny  mild     high  False   no

In [170]:
play_data[(play_data.outlook == 'sunny') & (play_data.humidity == 'normal')]

outlook  temp humidity  windy play
8    sunny  cool   normal  False  yes
10   sunny  mild   normal   True  yes

### Splitting the rainy branch

In [171]:
play_data[play_data.outlook == 'rainy']

outlook  temp humidity  windy play
3    rainy  mild     high  False  yes
4    rainy  cool   normal  False  yes
5    rainy  cool   normal   True   no
9    rainy  mild   normal  False  yes
13   rainy  mild     high   True   no

In [172]:
# Entropy(Play_Rainy|)
Entropy_Play_Outlook_Rainy =-(3/5)*np.log2(3/5) -(2/5)*np.log2(2/5)

In [173]:
Entropy_Play_Outlook_Rainy

0.9709505944546686

### Information Gain for temp
* mild -> 3 [2+ 1-]
* cool -> 2 [1+ 1-]

In [174]:
Entropy_Play_Outlook_Rainy - (3/5)*0.918 - (2/5)*1

0.020150594454668602

### Information Gain for Windy

In [175]:
Entropy_Play_Outlook_Rainy - (2/5)*0 - (3/5)*0

0.9709505944546686

### Information Gain for Humidity
* High -> 2 -> [1+ 1-]
* Normal -> 3 -> [2+ 1-]

In [177]:
Entropy_Play_Outlook_Rainy_Normal = -(1/3)*np.log2(1/3) - (2/3)*np.log2(2/3)

In [178]:
Entropy_Play_Outlook_Rainy_Normal

0.9182958340544896

In [179]:
Entropy_Play_Outlook_Rainy - (2/5)*1 - (3/5)*Entropy_Play_Outlook_Rainy_Normal

0.01997309402197489

### Final Tree

<img src="https://github.com/awantik/machine-learning-slides/blob/master/dt4.PNG?raw=true" width="600px">

In [184]:
play_data.dtypes

outlook     object
temp        object
humidity    object
windy         bool
play        object
dtype: object

In [183]:
play_data[(play_data.outlook == 'rainy') & (play_data.windy == 'False')]

C:\Users\BestintownACER1\Anaconda3\lib\site-packages\pandas\core\ops.py:1164: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


TypeError: invalid type comparison

In [180]:
##########################################################

# Decision Tree for Classification
* The leaf nodes of decision tree decides the class
* CART will convert features with continues values into categorical values
* Different tree will be generated with same data given in different order

In [ ]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier, export_graphviz, ExtraTreeClassifier

In [ ]:
iris = load_iris()

In [ ]:
iris

In [ ]:
iris.data[:5]

In [ ]:
iris.target[:5]

In [ ]:
dt = DecisionTreeClassifier(criterion='entropy')

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
trainX,testX,trainY,testY = train_test_split(iris.data, iris.target)

In [ ]:
dt.fit(trainX,trainY)

In [ ]:
export_graphviz(dt,'dt.tree')

### Visualizing the tree
* http://www.webgraphviz.com/

* Criteria - Entropy

In [ ]:
dt

In [ ]:
dt.predict(testX)

#### Feature Importances
* Important features will be higher up the tree
* We can use this techniques to identify important features

In [ ]:
dt.feature_importances_

#### Visualizing Decision Decision Boundry

In [ ]:
from sklearn.datasets import make_blobs

In [ ]:
X,Y = make_blobs(n_features=2, n_samples=1000, cluster_std=.8, centers=4, random_state=6)

In [ ]:
plt.scatter(X[:,0],X[:,1],c=Y,s=5, cmap='viridis')

In [ ]:
dt = DecisionTreeClassifier()

In [ ]:
dt.fit(X,Y)

In [ ]:
plot_step = 0.2
x_min, x_max = X[:, 0].min() - 1, X[:, 0].max() + 1
y_min, y_max = X[:, 1].min() - 1, X[:, 1].max() + 1
xx, yy = np.meshgrid(np.arange(x_min, x_max, plot_step),
                         np.arange(y_min, y_max, plot_step))

In [ ]:
outcome = dt.predict(np.c_[xx.ravel(), yy.ravel()])

In [ ]:
xx.shape

In [ ]:
plt.scatter(X[:,0],X[:,1],c=Y,s=5,cmap='viridis')
plt.scatter(xx.ravel(),yy.ravel(),c=outcome,s=1,alpha=1, cmap='viridis')
plt.show()

# Decision Tree for Regression
* Continues target is predicted with Tree.
* Decision Tree tries to partition data into subsets of homogenous contents ( minimize mean squared error )

In [ ]:
play_time = pd.read_csv('Data/tennis-time.csv.txt')

In [ ]:
play_time

In [ ]:
from sklearn.preprocessing import LabelEncoder
for col in ['outlook','temp','humidity','windy']:
    le = LabelEncoder()
    play_time[col] = le.fit_transform(play_time[col])

In [ ]:
from sklearn.tree import DecisionTreeRegressor

In [ ]:
dt = DecisionTreeRegressor()

In [ ]:
dt.fit(play_time.drop('time',axis=1), play_time.time)

In [ ]:
export_graphviz(dt,'regtree.dot',feature_names=['outlook','temp','humidity','windy'])

In [ ]:
play_time

#### View the tree from the Online website

*  http://www.webgraphviz.com/

In [ ]:
dt.predict([[2,1,1,1]])

In [ ]:
dt.feature_importances_

In [ ]:
# Advantages , Disadvantages & Pruning - To be continued